### ISSUE TITLE
Leveraging Numba as a Standalone Compiler like Nuitka without Intrusive Annotations

### ISSUE THREAD

**Issue by sirinath** [numba/numba#2630]
Having to add annotations to functiona is very intrusive. Sometimes you might be calling functions written by other parties and in libraries. It would be helpful to have Numba compile the code without annotations either AOT or using JIT.

**Comment by dhirschfeld** [comment]
Since a decorator is just a function which returns a function AFAICS you can easily  jit any function in your own code:
```python
from numba import jit
from thirdparty import slow_func

fast_func = jit(slow_func)
```

It's likely also possible to monkeypatch `thirdparty` to replace `slow_func` with the jitted version so that all internal functions will benefit.

*Disclaimer:* Just thinking out loud - I haven't tried it, YMMV.

**Comment by sirinath** [comment]
If a function calls another function which calls another functions ... These kind of usage becomes very cumbersome. Since most of the work is already done we can have an alternative python compiler which does AOT and JIT using Numba.

**Comment by dscole** [comment]
I completely agree with @sirinath. It seems silly that to make this work:
```
def some_func(var):
    return var + 5

@jit 
def my_func(var):
    return some_func(var)
```
 
you just need to add `@jit` to some_func, but numba can't provide that functionality for you. For our usage it would also be very helpful if while compiling a function and numba encounters an undecorated function it just tries to jit right there and then.

**Comment by seibert** [comment]
While I agree with the sentiment that Numba should try compiling dependent functions, Numba is sufficiently limited in what kinds of code it can compile that we prefer users to opt into compiling all needed functions for now.  It isn't clear how often compiling arbitrary functions in third-party code would actually work given the current limitations of the compiler.  (Maybe we've turned a corner on this that I'm not aware of...)

I think once object mode is a little more sophisticated (right now the only benefit is loop lifting) this is something we could investigate.  We would need to see if there is a way to mark functions that Numba previously failed to compile as such, so that we don't waste time attempting to compile them over and over.

**Comment by dscole** [comment]
In our use case, we provide our users with a library, and as part of it's api we accept user defined python functions. We want to be able to hide the fact that we use numba, as it's an implementation detail on our side. So what we do is we take the user defined function and "decorate" it ourselves (basically calling `jit(user_func)`). The problem is that it breaks down if the users split their implementation into more than one function, in which case they need to decorate all of them with `@jit`, which seems like unnecessary cruft. If there was at least a way for us to say `jit(user_func, recurse=True)` (and similarly `cfunc(..., recurse=True)`) that would have been great.

**Comment by sirinath** [comment]
Also being able to just run any Python program without additional decoration, imports or other special usage would be ideal. You take a valid python program and use alternate Python runtime based on Numba and everything runs faster with with JIT and also being able to generate standalone executables with AOT compiling.

**Comment by mcg1969** [comment]
@sirinath what would you consider to be Nuitka's primary limitations from your perspective? Presumably there is something (indeed, I hope a lot!) about Numba that compels you to want to use it despite the fact that it doesn't do many of the things that Nuitka does.

That said: +1 on facilitating the ability to `@jit` functions that include other byte-compiled (but non-jitted) functions therein.

**Comment by sirinath** [comment]
Numba has better corporate backing and more mature.

**Comment by mcg1969** [comment]
Thank you for the perspective!

**Comment by seibert** [comment]
OK, @dscole's UDF use case is a pretty compelling reason to make an opt-in opton like `@jit(callees=True)` (maybe a different name) that would automatically try to compile all called functions as well.  I've spun this idea out into a more focused issue: #2634.

Numba is definitely not designed for efficient whole program compilation at this point, so we won't be addressing that use case any time soon.

**Comment by esc** [comment]
Update on this  issue: #2634 was implemented by #4331. From what I understand, the request here initially was whole program compilation? Since the likelihood that we will be doing this anytime soon is low, I will close this issue now. If you would like this issue to be revisted, I would suggest initiating a conversation onabout this is the new Numba Discourse group at: https://numba.discourse.group/ -  perhaps we can revisit this at some point in Numba's future. Thank you for using Numba!
